File: app/models/startup.py
from pydantic import BaseModel, Field, field_validator
from typing import Literal

from app.models.scoring import StartupIntelligenceScore, PillarScoreBreakdown, EvidenceStatus
from app.models.evidence import Evidence
from app.models.analysis_context import AnalysisContext
from app.models.sps_v3 import SPSV3Assessment
from datetime import datetime


ConfidenceLevel = Literal["Low", "Medium", "High"]

# MVP hardening: bounds on oversized/empty input. Pydantic rejects a
# violating request with a 422 before the route body (and the real,
# multi-minute, paid pipeline call) ever runs. 50,000 characters is a
# generous ceiling for a company description -- comfortably above any
# real submission seen in testing (a few thousand characters) -- meant
# to stop a pathological paste (an entire scraped site, a whole PDF's
# raw text, etc.), not to constrain legitimate use. Pulled out to a
# constant (Unified Multi-Source Analyze Startup) so POST /analyze can
# reuse the exact same bound when validating its raw company_text form
# field via this same model, rather than duplicating the number.
MAX_COMPANY_TEXT_LENGTH = 50_000


class StartupAnalysisRequest(BaseModel):
    company_text: str = Field(min_length=1, max_length=MAX_COMPANY_TEXT_LENGTH)


class WebsiteAnalysisRequest(BaseModel):
    # Website / URL Ingestion: bounds obviously-invalid input with a fast
    # 422 before any network I/O -- the real security validation (scheme
    # allow-list, private-network rejection, DNS-rebinding-safe pinning,
    # bounded redirects/response size) happens in
    # app/website_scrapper.py, which is the one place that knowledge
    # should live. 2048 chars is a generous ceiling for a URL (well above
    # any real company website URL, including ones with query strings),
    # meant to reject a pathological paste, not to constrain real use.
    url: str = Field(min_length=1, max_length=2048)

    @field_validator("url")
    @classmethod
    def _must_look_like_a_url(cls, value: str) -> str:
        trimmed = value.strip()

        if not trimmed.lower().startswith(("http://", "https://")):
            raise ValueError("Website URL must start with http:// or https://")

        return trimmed


class PillarAnalysis(BaseModel):
    score: float | None = None
    confidence: ConfidenceLevel = "Low"
    summary: str = ""
    evidence: list[Evidence | str] = Field(default_factory=list)
    strengths: list[str] = Field(default_factory=list)
    weaknesses: list[str] = Field(default_factory=list)
    recommendations: list[str] = Field(default_factory=list)
    score_breakdown: PillarScoreBreakdown = Field(default_factory=PillarScoreBreakdown)


class SIEContext(BaseModel):
    company_name: str = ""
    industry: str = ""
    business_model: str = ""
    company_stage: str = ""
    funding_stage: str = ""


class PartialStructuralCoverage(BaseModel):
    """
    SIE Methodology v2, Part 9 item 6 (Blocker 3 fix, post-implementation
    review): a purely additive, DISPLAY-ONLY label for whole-pillar evidence
    absence. Computed by
    app.ai.sie_v2_evidence_semantics.compute_partial_structural_coverage()
    from each pillar's already-existing final score (None == that pillar had
    zero scored dimensions). Never a math adjustment to SPS -- startup_
    intelligence_score is computed identically whether or not this field is
    populated. field is None (not this model's zero-value) on every analysis
    that predates this field, so historical JSONB records are never silently
    reinterpreted as "fully covered."
    """

    partial_structural_coverage: bool = False
    pillars_unavailable_entirely: list[str] = Field(default_factory=list)
    note: str = ""


class SIEMethodologyAnalysis(BaseModel):
    context: SIEContext = Field(default_factory=SIEContext)

    analysis_context: AnalysisContext | None = None

    market: PillarAnalysis = Field(default_factory=PillarAnalysis)
    team: PillarAnalysis = Field(default_factory=PillarAnalysis)
    product: PillarAnalysis = Field(default_factory=PillarAnalysis)
    execution: PillarAnalysis = Field(default_factory=PillarAnalysis)
    traction: PillarAnalysis = Field(default_factory=PillarAnalysis)
    financial_health: PillarAnalysis = Field(default_factory=PillarAnalysis)

    startup_intelligence_score: float = 0.0
    startup_scorecard: StartupIntelligenceScore = Field(default_factory=StartupIntelligenceScore)

    milestone_readiness_score: float = 0.0
    momentum_score: float = 0.0
    confidence_score: float = 0.0

    # None (not PartialStructuralCoverage()'s own zero-value) is the
    # backward-compatible default: a record stored before this field existed
    # decodes to None here, never to a fabricated "fully covered" reading.
    structural_coverage: PartialStructuralCoverage | None = None

    executive_coaching_summary: str = ""
    next_actions: list[str] = Field(default_factory=list)

    # Phase 10.9, Part 5: the SPS V3 assessment, additive alongside every
    # V2.1 field above (startup_intelligence_score, startup_scorecard,
    # etc.) -- none of which this touches or reinterprets. None on every
    # analysis produced before this field existed, and on every analysis
    # produced while the V3 feature flag is off (see
    # app/ai/sps_v3_adapter.py's SPS_V3_ENABLED) -- never backfilled,
    # never inferred from the V2.1 fields. A non-None value here is the
    # ONLY signal that this analysis has a V3 assessment at all.
    sps_v3: SPSV3Assessment | None = None



class StartupAnalysisResponse(BaseModel):
    context: SIEContext = Field(default_factory=SIEContext)
    startup_scorecard: StartupIntelligenceScore = Field(default_factory=StartupIntelligenceScore)
    methodology: SIEMethodologyAnalysis = Field(default_factory=SIEMethodologyAnalysis)


class UpdateAnalysisRequest(BaseModel):
    methodology: SIEMethodologyAnalysis

class StartupProfileResponse(BaseModel):
    id: int
    # Saved Startups (Watchlist Phase 1): the canonical Startup FK
    # (analyses.startup_id), additive alongside the existing analysis-id
    # `id` field above -- see get_startup_by_name()'s docstring. None only
    # for historical rows that predate both the write path and its
    # backfill; the frontend Save control hides itself when this is None
    # rather than guessing an id to save.
    startup_id: int | None = None
    # Phase 37E -- Company Lifecycle + Public Identity Convergence,
    # Section 6. The company's own canonical name -- always present, even
    # when has_analysis is False, so the honest "not yet evaluated" state
    # can still say whose page this is.
    canonical_name: str
    created_at: datetime
    # None exactly when has_analysis is False -- a company that exists
    # (a real `startups` row) but has never been analyzed. Never a
    # fabricated SPS/pillar breakdown; the frontend renders a distinct,
    # honest state for this case instead of treating it as "not found."
    methodology: SIEMethodologyAnalysis | None = None
    has_analysis: bool = True


class SavedStartupEntry(BaseModel):
    """
    Saved Startups (Watchlist Phase 1) list-row shape -- see
    get_saved_startups_for_user()'s docstring. Deliberately flat and
    minimal (mirrors RankingEntry's shape, not the full nested
    SIEMethodologyAnalysis) since this is a list view, not a second
    startup-details experience; every field here is read fresh from the
    startup's current latest canonical analysis on every request, never
    copied into saved_startups itself. industry/stage/overall_score/
    latest_analysis_at are None when the saved startup currently has no
    canonical analysis -- never fabricated.
    """
    startup_id: int
    company_name: str
    industry: str | None = None
    stage: str | None = None
    overall_score: float | None = None
    latest_analysis_at: datetime | None = None
    saved_at: datetime


class MyAnalysisEntry(BaseModel):
    """
    Portfolio Release Task 4 -- My Analyses list-row shape. See
    get_my_analyses()'s own docstring for why this is one row PER
    ANALYSIS (submitted_by_user_id = the caller), not one row per
    startup the way SavedStartupEntry above is -- two users (or the same
    user, twice) analyzing the same company_name are two separate rows
    here, each reopening its own submitter's own authorized report.
    startup_id/company_name are nullable only for the same defensive
    reason StartupProfileResponse's own startup_id is (a row whose
    startup somehow doesn't resolve) -- never expected in practice, since
    every write path that sets submitted_by_user_id also always resolves
    a real startup_id (see save_analysis()'s own docstring), but the
    frontend still guards against it rather than assuming.
    """
    analysis_id: int
    startup_id: int | None = None
    company_name: str | None = None
    overall_score: float | None = None
    created_at: datetime


class SavedStartupStatus(BaseModel):
    saved: bool


class DiscoveryResult(BaseModel):
    """
    Startup Discovery V1 row shape -- deliberately flat, mirroring
    SavedStartupEntry/RankingEntry (a list view, not a second
    startup-details experience). Every field is read fresh from that
    startup's current latest canonical analysis on every request -- see
    discover_startups()'s own docstring in app/database/db.py. Pillar
    fields are None when that pillar was Unavailable on the underlying
    analysis, never fabricated.
    """
    startup_id: int
    company_name: str
    industry: str | None = None
    stage: str | None = None
    business_model: str | None = None
    overall_score: float | None = None
    market_score: float | None = None
    team_score: float | None = None
    product_score: float | None = None
    execution_score: float | None = None
    traction_score: float | None = None
    financial_score: float | None = None
    created_at: datetime


class DiscoveryResponse(BaseModel):
    total: int
    results: list[DiscoveryResult]


class DiscoveryFilterOptions(BaseModel):
    """
    Startup Discovery V1, Part 4: option lists derived from the real
    canonical population -- see get_discovery_filter_options()'s own
    docstring. Never a hardcoded taxonomy.
    """
    industries: list[str]
    stages: list[str]
    business_models: list[str]


# ---------------------------------------------------------------------------
# Compare Startups V1. This is a deliberately slimmer, purpose-built set of
# models -- not a reuse of PillarAnalysis/SIEMethodologyAnalysis wholesale
# -- so the /compare payload carries only what a side-by-side comparison
# actually needs (no raw evidence quotes, no executive_coaching_summary,
# no momentum/confidence scores never populated by the backend). Every
# field here is read directly from the existing stored methodology JSONB;
# nothing is recomputed, re-scored, or re-summarized by an LLM.
# ---------------------------------------------------------------------------

class ComparisonSubscore(BaseModel):
    name: str
    # None means this dimension could not be responsibly scored --
    # preserved as None here too, never coerced to 0.
    score: float | None = None
    weight: float = 0.0
    confidence: ConfidenceLevel = "Low"
    evidence_status: EvidenceStatus = "Observed"
    rationale: str = ""
    recommendations: list[str] = Field(default_factory=list)
    missing_information: list[str] = Field(default_factory=list)


class ComparisonPillar(BaseModel):
    pillar: str
    # None means no dimensions in this pillar were scorable -- i.e. this
    # pillar is Unavailable for this startup. Never defaulted to 0.
    score: float | None = None
    confidence: ConfidenceLevel = "Low"
    evidence_coverage: float = 0.0
    summary: str = ""
    strengths: list[str] = Field(default_factory=list)
    weaknesses: list[str] = Field(default_factory=list)
    recommendations: list[str] = Field(default_factory=list)
    subscores: list[ComparisonSubscore] = Field(default_factory=list)


class ComparisonStartup(BaseModel):
    startup_id: int
    company_name: str
    industry: str = ""
    company_stage: str = ""
    business_model: str = ""
    latest_analysis_at: datetime
    overall_score: float | None = None

    market: ComparisonPillar
    team: ComparisonPillar
    product: ComparisonPillar
    execution: ComparisonPillar
    traction: ComparisonPillar
    financial_health: ComparisonPillar

    # Phase 10.9, Part 21 -- additive passthrough only. None whenever this
    # startup's latest analysis has no sps_v3 (every historical analysis,
    # and every analysis produced while the V3 feature flag is off).
    # Compare never manufactures comparability between a V2.1-only
    # startup and a V3-assessed one -- see SPS_V3_PRODUCTION_INTEGRATION_10_9.md
    # Section 21 for the exact UI contract this enables.
    sps_v3: SPSV3Assessment | None = None


class ComparisonResponse(BaseModel):
    startups: list[ComparisonStartup]
    # Requested startup_ids that could not be resolved to a canonical
    # analysis -- an invalid/nonexistent id, or a real startup whose only
    # analyses predate Methodology v2. Never silently dropped without a
    # trace: the frontend surfaces this honestly rather than pretending
    # the request was fully satisfied.
    missing_startup_ids: list[int] = Field(default_factory=list)

File: dashboard/types/startup.ts
// Canonical Dashboard MVP: get_top_startups() now returns exactly
// get_rankings()'s rows (see app/database/db.py), which never include
// readiness_score -- that field was dropped, not just left unrendered
// (readiness_score has no defined numeric scale; see the P0 Product Trust
// Cleanup report).
export interface StartupRanking {
  company_name: string;
  industry: string;
  stage: string;
  overall_score: number;
}

export interface ImprovingStartup {
  company_name: string;
  first_score: number;
  latest_score: number;
  score_change: number;
}

export interface RankingEntry {
  id: number;
  company_name: string | null;
  industry: string | null;
  stage: string | null;
  business_model: string | null;
  overall_score: number | null;
  market_score: number | null;
  team_score: number | null;
  product_score: number | null;
  competition_score: number | null;
  traction_score: number | null;
  financial_score: number | null;
  recommendation: string | null;
  created_at: string;
}

export type ConfidenceLevel = "Low" | "Medium" | "High";

export type Evidence = {
  source?: string;
  text?: string;
  title?: string;
  url?: string;
};

export type Subscore = {
  name: string;
  score: number | null;
  weight: number;
  confidence?: ConfidenceLevel;
  evidence_status?: string;
  rationale?: string;
  evidence?: string[];
  recommendations?: string[];
  missing_information?: string[];
};

export type PillarScoreBreakdown = {
  pillar?: string;
  score?: number | null;
  confidence?: ConfidenceLevel;
  evidence_coverage?: number;
  scoring_summary?: string;
  subscores?: Subscore[];
};

export type PillarAnalysis = {
  score: number | null;
  confidence: ConfidenceLevel;
  summary: string;
  evidence: Array<Evidence | string>;
  strengths: string[];
  weaknesses: string[];
  recommendations: string[];
  score_breakdown: PillarScoreBreakdown;
};

export type SIEContext = {
  company_name: string;
  industry: string;
  business_model: string;
  company_stage: string;
  funding_stage: string;
};

// SIE Methodology v2, Part 9 item 6 (Blocker 3): a purely additive,
// display-only whole-pillar-coverage label -- never a math adjustment to
// startup_intelligence_score. Optional/nullable because analyses stored
// before this field existed have no key for it at all.
export type PartialStructuralCoverage = {
  partial_structural_coverage: boolean;
  pillars_unavailable_entirely: string[];
  note: string;
};

// Phase 10.9, Part 5: mirrors app/models/sps_v3.py's SPSV3Assessment /
// SPSV3PillarResult. Additive alongside every V2.1 field on
// SIEMethodologyAnalysis below -- undefined/null on any analysis that
// predates this field, or that was produced while the backend's V3
// feature flag was off. A non-null value here is the ONLY signal that
// this analysis has a V3 assessment at all -- see
// docs/methodology/SPS_V3_PRODUCTION_INTEGRATION_10_9.md.
export type SPSV3ConfidenceLevel = "Low" | "Medium" | "High";

export type SPSV3AssessmentState = "sufficient" | "limited" | "insufficient";

export type SPSV3PillarResult = {
  pillar: string;
  // null means this pillar itself did not clear its own coverage floor
  // -- never a fabricated/defaulted number.
  strength: number | null;
  coverage_pct: number;
  confidence: SPSV3ConfidenceLevel;
  publishable: boolean;
  withhold_reason: string | null;
};

export type SPSV3Assessment = {
  engine_version: string;
  scoring_version: string;
  // The published Startup Power Score under the V3 methodology. null is
  // a legitimate, common, EXPECTED value (assessment_state !==
  // "sufficient") -- it must never be read as, or rendered as, 0.
  overall_score: number | null;
  coverage_pct: number;
  confidence: SPSV3ConfidenceLevel;
  assessment_state: SPSV3AssessmentState;
  withhold_reason: string | null;
  pillars: Record<string, SPSV3PillarResult>;
  computed_at: string;
};

// Task 5 -- Unified Visual Design (AI transparency addendum). Mirrors
// app/models/analysis_context.py::AnalysisContext -- every field is
// optional/defaulted there too ("Populated only for analyses run after
// this field set was added... loading an older stored methodology JSONB
// -- which never had these keys -- leaves them blank rather than
// pretending that analysis carried provenance it never actually
// recorded"). Replaces the previous `analysis_context?: unknown` so the
// new "How this analysis was generated" section can read these fields
// safely, typed -- this is a frontend typing addition only, not a
// backend contract change; the JSON shape was already exactly this.
export type SIEAnalysisContext = {
  analysis_type?: string;
  evidence_sources?: string[];
  missing_information?: string[];
  methodology_version?: string;
  anchor_registry_version?: string;
  scoring_version?: string;
  model_identifier?: string;
  prompt_version?: string;
  company_text_hash?: string;
  search_query?: string;
  research_brief_snapshot?: string;
  source_snapshot?: Array<{ title?: string; url?: string }>;
  analyzed_at?: string;
};

export type SIEMethodologyAnalysis = {
  context: SIEContext;

  market: PillarAnalysis;
  team: PillarAnalysis;
  product: PillarAnalysis;
  execution: PillarAnalysis;
  traction: PillarAnalysis;
  financial_health: PillarAnalysis;

  startup_intelligence_score: number;
  milestone_readiness_score: number;
  momentum_score: number;
  confidence_score: number;

  executive_coaching_summary: string;
  next_actions: string[];

  structural_coverage?: PartialStructuralCoverage | null;

  startup_scorecard?: unknown;
  analysis_context?: SIEAnalysisContext;

  sps_v3?: SPSV3Assessment | null;
};

export type StartupProfileResponse = {
  id: number;
  // Saved Startups (Watchlist Phase 1): the canonical Startup FK, additive
  // alongside `id` (which is the analysis id, not the startup id -- see
  // the backend's get_startup_by_name() docstring). null only for
  // historical rows that predate the write path; the Save control hides
  // itself rather than guessing an id to save.
  startup_id: number | null;
  // Phase 37E: always present, even when has_analysis is false.
  canonical_name: string;
  created_at: string;
  // Phase 37E -- Company Lifecycle + Public Identity Convergence,
  // Section 6: null exactly when has_analysis is false -- a company that
  // exists but hasn't been analyzed yet. Never fabricated.
  methodology: SIEMethodologyAnalysis | null;
  has_analysis: boolean;
};

// Saved Startups (Watchlist Phase 1): GET /me/saved-startups row shape.
// Deliberately flat, mirroring RankingEntry -- every field is read fresh
// from the startup's current latest canonical analysis on every request,
// never a snapshot copied in at save time. industry/stage/overall_score/
// latest_analysis_at are null when the saved startup currently has no
// canonical analysis -- never fabricated.
export type SavedStartupEntry = {
  startup_id: number;
  company_name: string;
  industry: string | null;
  stage: string | null;
  overall_score: number | null;
  latest_analysis_at: string | null;
  saved_at: string;
};

export type SavedStartupStatus = {
  saved: boolean;
};

// Portfolio Release Task 4 -- My Analyses: GET /me/analyses row shape.
// One row PER ANALYSIS (submitted_by_user_id = the caller), not per
// startup -- see app/models/startup.py::MyAnalysisEntry's own docstring
// for why this differs from SavedStartupEntry above. startup_id/
// company_name are nullable only defensively; every real write path
// that sets submitted_by_user_id also resolves a real startup_id.
export type MyAnalysisEntry = {
  analysis_id: number;
  startup_id: number | null;
  company_name: string | null;
  overall_score: number | null;
  created_at: string;
};

// One point per canonical (methodology-bearing) analysis, sourced from
// GET /startup/{company_name}/sps-history. Chronological order.
export type SPSHistoryPoint = {
  analysis_id: number;
  created_at: string;
  startup_intelligence_score: number;
};
